### ISSUE TITLE
pipe doesn't seem to close when process on the other end is killed

### ISSUE THREAD

**Issue by zaddach** [oconnor663/os_pipe.rs#20]
I'm trying to write a program that consumes output from a process, and terminates the process if a timeout is reached. I've managed to get the output with `os_pipe` and kill the process with `shared_child`, but my program doesn't terminate.

Cargo.toml:
```
[package]
name = "subproc"
version = "0.1.0"
authors = ["<snip>"]
edition = "2018"

[dependencies]
os_pipe = "0.9.2"
```
src/main.rs:
```rust
use std::process::Command;
use std::io::{BufRead, BufReader};
use std::thread;
use std::time;
use os_pipe::pipe;

fn main() {
    const TIMEOUT : i32 = 5;
    let (stdout_read, stdout_write) = pipe().unwrap();
    let mut cmd = Command::new("bash");
    cmd.arg("-c").arg("for ((i=1000; i > 0; i--)); do echo \"$i bottles of beer on the wall\"; sleep 1; done");
    cmd.stdout(stdout_write);
    let mut child = cmd.spawn().unwrap();

    let reader = BufReader::new(stdout_read);

    let _thread = thread::spawn(move || {
        for _ in 0..TIMEOUT {
            if let Ok(Some(_)) = child.try_wait() {
                return;
            }
            thread::sleep(time::Duration::from_secs(1));
        }

        child.kill().unwrap();
        drop(child);
    });


    reader
        .lines()
        .filter_map(|line| line.ok())
        .for_each(|line| println!("{}", line));

    println!("Hello, world!");
}
```

Expected output:
```
1000 bottles of beer on the wall
999 bottles of beer on the wall
998 bottles of beer on the wall
997 bottles of beer on the wall
996 bottles of beer on the wall
Hello, world!
```
and the rust program terminates.

Actual output:
```
1000 bottles of beer on the wall
999 bottles of beer on the wall
998 bottles of beer on the wall
997 bottles of beer on the wall
996 bottles of beer on the wall
```
and the program doesn't terminate.

Thanks for any help!

**Comment by oconnor663** [comment]
You're running into the same issue that this comment is talking about in the [README example](https://github.com/oconnor663/os_pipe.rs#example):

```
// Very important when using pipes: This parent process is still
// holding its copies of the write ends, and we have to close them
// before we read, otherwise the read end will never report EOF. The
// Command object owns the writers now, and dropping it closes them.
```

Your reader is never going to see EOF, because your `Command` object is still holding an open write pipe. If you `drop(cmd);` before you try to read, it'll get rid of the deadlock.

The `duct` crate is intended to take care of most of these details in common cases, without any risk of deadlocks like this. Maybe give it a look. The `std::process::Stdio::piped()` API is also somewhat less likely to cause this bug, because it opens pipes during spawning, rather than holding them in the `Command` itself.

**Comment by oconnor663** [comment]
That said, there's another hazard I want to point out here, which isn't a problem yet but which could become one. There's a limitation in how `Command::kill` works, which makes it less effective than Ctrl-C in the shell. The shell is able to signal entire "process groups" at once. But for [unfortunate historical Unix reasons](https://github.com/oconnor663/duct.py/blob/master/gotchas.md#killing-grandchild-processes), the `Command::kill` method (along with most similar library functions in most languages, as far as I know) can only signal the specific process you spawned. This is particularly important when you're invoking `bash -c ...`, because it means you're going to kill Bash, but you're not going to kill any of the grandchild processes that you're telling Bash to spawn for you. That's not the end of the world in your example code, because the `sleep` process is going to exit on its own in 1 second. But if you were sleeping for longer, it could become a problem. (And if you're noticing a weird extra second of lag before your program exits, this is likely why.) Note that the orphaned `sleep` process won't block you when you wait on Bash to exit, but it _will_ block you waiting on EOF from the stdout pipe, because it will have inherited a duplicate of that pipe.

**Comment by oconnor663** [comment]
I'm going to close this for now, but feel free to comment with more questions if any of that is unclear.

**Comment by zaddach** [comment]
Hey, thanks a lot for the help. I got my example working even without the `os_pipe` crate, I'll post the code here in case somebody else stumbles upon this issue. And I'll keep your warning about killing the process vs a process group in mind - in my case (apart from the example I posted here) I don't expect it to cause a problem.

Working code:
```rust
use std::process::{Command, Stdio};
use std::io::{BufRead, BufReader};
use std::thread;
use std::time;

fn main() {
    const TIMEOUT : i32 = 5;
    let mut cmd = Command::new("bash");
    cmd.arg("-c").arg("for ((i=1000; i > 0; i--)); do echo \"$i bottles of beer on the wall\"; sleep 1; done");
    cmd.stdout(Stdio::piped());
    let mut child = cmd.spawn().unwrap();
    let stdout = child.stdout.take().unwrap();


    let thread = thread::spawn(move || {
        for _ in 0..TIMEOUT {
            if let Ok(Some(_)) = child.try_wait() {
                return;
            }
            thread::sleep(time::Duration::from_secs(1));
        }

        child.kill().unwrap();
    });

    let reader = BufReader::new(stdout);
    for line in reader.lines() {
        println!("line: {}", line.unwrap());
    }
    thread.join().unwrap();

    println!("Hello, world!");
}
```
